packages/@aws-cdk/aws-apigateway/lib/integrations/stepfunctions.ts

```diff
@@ -0,0 +1,237 @@
+    } else {
+      //imported state machine
+      stateMachineName = 'StateMachine-' + (String(this.stateMachine.stack.node.addr));
+    }
+
+    let deploymentToken;
+
+    if (stateMachineName !== undefined && !Token.isUnresolved(stateMachineName)) {
+      deploymentToken = JSON.stringify({ stateMachineName });
+    }
+    return {
+      ...bindResult,
+      deploymentToken,
+    };
+
+  }
+}
+
+function integrationResponse() {
+  const errorResponse = [
+    {
+      /**
+       * Specifies the regular expression (regex) pattern used to choose
+       * an integration response based on the response from the back end.
+       * In this case it will match all '4XX' HTTP Errors
+       */
+      selectionPattern: '4\\d{2}',
+      statusCode: '400',
+      responseTemplates: {
+        'application/json': `{
+            "error": "Bad input!"
+          }`,
+      },
+    },
+    {
+      /**
+       * Match all '5XX' HTTP Errors
+       */
+      selectionPattern: '5\\d{2}',
+      statusCode: '500',
+      responseTemplates: {
+        'application/json': '"error": $input.path(\'$.error\')',
+      },
+    },
+  ];
+
+  const integResponse = [
+    {
+      statusCode: '200',
+      responseTemplates: {
+        'application/json': `#set($inputRoot = $input.path('$'))
+                #if($input.path('$.status').toString().equals("FAILED"))
+                    #set($context.responseOverride.status = 500)
+                    { 
+                      "error": "$input.path('$.error')",
+                      "cause": "$input.path('$.cause')"
+                    }
+                #else
+                    $input.path('$.output')
+                #end`,
+      },
+    },
+    ...errorResponse,
+  ];
+
+  return integResponse;
+}
+
+function requestTemplates(stateMachine: sfn.IStateMachine, includeRequestContext: boolean, requestContextObj: RequestContext | undefined) {
+  const templateStr = templateString(stateMachine, includeRequestContext, requestContextObj);
+
+  const requestTemplate: { [contentType:string] : string } =
+    {
+      'application/json': templateStr,
+    };
+
+  return requestTemplate;
+}
+
+function templateString(stateMachine: sfn.IStateMachine, includeRequestContext: boolean, requestContextObj: RequestContext | undefined): string {
+  let templateStr: string;
+
+  if (includeRequestContext) {
+    const requestContextStr = requestContext(requestContextObj);
+
+    const search = '"';
+    const replaceWith = '\\"';
+    const requestContextStrModified = requestContextStr.split(search).join(replaceWith);
+    templateStr = `
+    #set($allParams = $input.params())
+    {
+      "input": "{${requestContextStrModified}}",
+      "stateMachineArn": "${stateMachine.stateMachineArn}"
+    }`;
+  } else {
+    templateStr = `
+    #set($inputRoot = $input.path('$')) {
+        "input": "$util.escapeJavaScript($input.json('$'))",
```

Comment:
could you explain the difference between this if-else. Why use `escapeJavascript()` in the `else` case?

Reply:
if request context is specified, it will go into the if case, otherwise it will go into the else case. The `escapeJavascript()` escapes the characters in a string using JavaScript string rules (more at https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-mapping-template-reference.html).

Reply:
The parameters passed are always `input` and `stateMachineArn`.  When request Context is specified, the request payload along with `requestContext` object is passed into the `input` key. The request payload is passed with the `body` key inside the `input` key and `requestContext` is passed with the `requestContext` key inside the `input` key.

Reply:
Feel very roundabout, and hard for users to reason about.

Why not have the input passed into the `input` key all the time, and a `requestContext` key with the request context?